File: src/agent_system/memory.py
from __future__ import annotations

import hashlib
import math
import re
from datetime import UTC, datetime

from sqlalchemy import select

from .db import Database
from .models import (
    MemoryCandidateRecord,
    MemoryRecord,
    MemorySourceRecord,
    RunRecord,
)

TOKEN_PATTERN = re.compile(r"[\w\u4e00-\u9fff]+", re.UNICODE)
SECRET_PATTERN = re.compile(
    r"(?:sk-|tp-)[A-Za-z0-9_-]{20,}|(?:api[_-]?key|token|secret)\s*[:=]\s*\S+",
    re.IGNORECASE,
)


def tokenize(text: str) -> list[str]:
    return [token.lower() for token in TOKEN_PATTERN.findall(text)]


def hash_embedding(text: str, dimensions: int = 128) -> list[float]:
    vector = [0.0] * dimensions
    for token in tokenize(text):
        digest = hashlib.blake2b(token.encode("utf-8"), digest_size=8).digest()
        index = int.from_bytes(digest, "big") % dimensions
        vector[index] += 1.0
    norm = math.sqrt(sum(value * value for value in vector)) or 1.0
    return [value / norm for value in vector]


def cosine(left: list[float], right: list[float]) -> float:
    return sum(a * b for a, b in zip(left, right, strict=False))


class MemoryService:
    def __init__(self, db: Database):
        self.db = db

    def _validate_content(self, content: str) -> None:
        if SECRET_PATTERN.search(content):
            raise ValueError("memory content appears to contain a credential")

    async def create(
        self,
        *,
        tenant_id: str,
        scope_type: str,
        scope_id: str,
        kind: str,
        content: str,
        confidence: float,
        source: dict | None = None,
        run_id: str | None = None,
    ) -> MemoryRecord:
        self._validate_content(content)
        digest = hashlib.sha256(content.strip().encode("utf-8")).hexdigest()
        async with self.db.sessions() as session, session.begin():
            existing = await session.scalar(
                select(MemoryRecord).where(
                    MemoryRecord.tenant_id == tenant_id,
                    MemoryRecord.scope_type == scope_type,
                    MemoryRecord.scope_id == scope_id,
                    MemoryRecord.content_hash == digest,
                    MemoryRecord.enabled.is_(True),
                )
            )
            if existing is not None:
                existing.confidence = max(existing.confidence, confidence)
                return existing
            embedding = hash_embedding(content)
            memory = MemoryRecord(
                tenant_id=tenant_id,
                scope_type=scope_type,
                scope_id=scope_id,
                kind=kind,
                content=content.strip(),
                content_hash=digest,
                embedding_json=embedding,
                embedding_vector=embedding,
                confidence=confidence,
            )
            session.add(memory)
            await session.flush()
            session.add(
                MemorySourceRecord(memory_id=memory.id, run_id=run_id, source_json=source or {})
            )
        return memory

    async def create_candidate_from_run(self, run_id: str) -> MemoryCandidateRecord | None:
        async with self.db.sessions() as session:
            run = await session.get(RunRecord, run_id)
            if run is None or run.status != "completed" or run.output_json is None:
                return None
        memory_scope = run.context_json.get("memory_scope") or {}
        scope_type = memory_scope.get("type", "agent")
        scope_id = memory_scope.get("id", run.agent_version_id)
        content = f"Input: {str(run.input_json)[:2000]}\nOutcome: {str(run.output_json)[:4000]}"
        if SECRET_PATTERN.search(content):
            return None
        candidate = MemoryCandidateRecord(
            tenant_id=run.tenant_id,
            run_id=run.id,
            scope_type=scope_type,
            scope_id=scope_id,
            kind="episodic",
            content=content,
            scores_json={"usefulness": 0.6, "confidence": 0.6, "safety": 1.0},
        )
        async with self.db.sessions() as session, session.begin():
            session.add(candidate)
        return candidate

    async def consolidate(self, candidate_id: str) -> MemoryRecord:
        async with self.db.sessions() as session, session.begin():
            candidate = await session.scalar(
                select(MemoryCandidateRecord)
                .where(MemoryCandidateRecord.id == candidate_id)
                .with_for_update()
            )
            if candidate is None:
                raise KeyError("memory candidate not found")
            if candidate.status == "accepted":
                source = await session.scalar(
                    select(MemorySourceRecord).where(MemorySourceRecord.run_id == candidate.run_id)
                )
                if source:
                    memory = await session.get(MemoryRecord, source.memory_id)
                    if memory:
                        return memory
            candidate.status = "processing"
        try:
            memory = await self.create(
                tenant_id=candidate.tenant_id,
                scope_type=candidate.scope_type,
                scope_id=candidate.scope_id,
                kind=candidate.kind,
                content=candidate.content,
                confidence=float(candidate.scores_json.get("confidence", 0.5)),
                source={"candidate_id": candidate.id},
                run_id=candidate.run_id,
            )
        except Exception:
            async with self.db.sessions() as session, session.begin():
                current = await session.get(MemoryCandidateRecord, candidate_id)
                if current:
                    current.status = "rejected"
            raise
        async with self.db.sessions() as session, session.begin():
            current = await session.get(MemoryCandidateRecord, candidate_id)
            if current:
                current.status = "accepted"
        return memory

    async def retrieve(
        self,
        *,
        tenant_id: str,
        query: str,
        scopes: list[tuple[str, str]],
        limit: int = 8,
    ) -> list[dict]:
        if not scopes:
            return []
        now = datetime.now(UTC)
        async with self.db.sessions() as session:
            rows = (
                await session.scalars(
                    select(MemoryRecord).where(
                        MemoryRecord.tenant_id == tenant_id,
                        MemoryRecord.enabled.is_(True),
                    )
                )
            ).all()
        allowed = set(scopes)
        query_tokens = set(tokenize(query))
        query_vector = hash_embedding(query)
        ranked = []
        for row in rows:
            if (row.scope_type, row.scope_id) not in allowed:
                continue
            valid_to = row.valid_to
            if valid_to is not None:
                if valid_to.tzinfo is None:
                    valid_to = valid_to.replace(tzinfo=UTC)
                if valid_to <= now:
                    continue
            memory_tokens = set(tokenize(row.content))
            keyword = len(query_tokens & memory_tokens) / max(len(query_tokens), 1)
            semantic = cosine(query_vector, list(row.embedding_json or []))
            score = 0.55 * semantic + 0.25 * keyword + 0.20 * row.confidence
            ranked.append((score, row))
        ranked.sort(key=lambda item: item[0], reverse=True)
        return [
            {
                "id": row.id,
                "kind": row.kind,
                "content": row.content,
                "confidence": row.confidence,
                "score": round(score, 6),
                "updated_at": row.updated_at.isoformat(),
            }
            for score, row in ranked[:limit]
        ]

File: src/memory.ts
import { createHash } from "node:crypto";
import type { Database } from "./database.js";
import { NotFoundError, required } from "./database.js";
import { checksum } from "./crypto.js";
import { newId } from "./serialization.js";

interface MemoryRow extends Record<string, unknown> {
  id: string;
  tenant_id: string;
  scope_type: string;
  scope_id: string;
  kind: string;
  content: string;
  content_hash: string;
  embedding_json: number[];
  confidence: number;
  enabled: boolean;
  valid_from: string;
  valid_to: string | null;
  superseded_by: string | null;
  source_json: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export function hashEmbedding(text: string, dimensions = 128): number[] {
  const bytes = createHash("sha256").update(text.toLowerCase()).digest();
  const values = Array.from(
    { length: dimensions },
    (_, i) => bytes[i % bytes.length]! / 127.5 - 1,
  );
  const norm = Math.sqrt(values.reduce((sum, v) => sum + v * v, 0)) || 1;
  return values.map((v) => v / norm);
}
const cosine = (a: number[], b: number[]) =>
  a.reduce((sum, value, index) => sum + value * (b[index] ?? 0), 0);

export class MemoryService {
  constructor(private readonly db: Database) {}
  async create(
    tenantId: string,
    input: Record<string, unknown>,
    source: Record<string, unknown> = {},
  ) {
    const content = String(input.content);
    const id = newId();
    return required<MemoryRow>(
      this.db,
      `INSERT INTO memories(id,tenant_id,scope_type,scope_id,kind,content,content_hash,embedding_json,confidence,source_json)
    VALUES($1,$2,$3,$4,$5,$6,$7,$8::jsonb,$9,$10::jsonb) RETURNING *`,
      [
        id,
        tenantId,
        String(input.scope_type),
        String(input.scope_id),
        String(input.kind ?? "semantic"),
        content,
        checksum(content),
        JSON.stringify(hashEmbedding(content)),
        Number(input.confidence ?? 0.8),
        JSON.stringify(source),
      ],
    );
  }
  async list(tenantId: string, filters: Record<string, unknown> = {}) {
    const params: unknown[] = [tenantId];
    const conditions = ["tenant_id=$1"];
    for (const key of ["scope_type", "scope_id", "kind"]) {
      if (filters[key] !== undefined) {
        params.push(filters[key]);
        conditions.push(`${key}=$${params.length}`);
      }
    }
    if (filters.enabled !== undefined) {
      params.push(Boolean(filters.enabled));
      conditions.push(`enabled=$${params.length}`);
    }
    params.push(Math.min(Number(filters.limit ?? 200), 1000));
    return (
      await this.db.query<MemoryRow>(
        `SELECT * FROM memories WHERE ${conditions.join(" AND ")} ORDER BY updated_at DESC LIMIT $${params.length}`,
        params,
      )
    ).rows;
  }
  async patch(tenantId: string, id: string, input: Record<string, unknown>) {
    const current = await required<MemoryRow>(
      this.db,
      "SELECT * FROM memories WHERE id=$1 AND tenant_id=$2",
      [id, tenantId],
      "memory not found",
    );
    const content =
      input.content === undefined ? current.content : String(input.content);
    return required<MemoryRow>(
      this.db,
      `UPDATE memories SET content=$3,content_hash=$4,embedding_json=$5::jsonb,confidence=$6,enabled=$7,updated_at=now() WHERE id=$1 AND tenant_id=$2 RETURNING *`,
      [
        id,
        tenantId,
        content,
        checksum(content),
        JSON.stringify(hashEmbedding(content)),
        input.confidence === undefined
          ? current.confidence
          : Number(input.confidence),
        input.enabled === undefined ? current.enabled : Boolean(input.enabled),
      ],
    );
  }
  async remove(tenantId: string, id: string) {
    const result = await this.db.query(
      "DELETE FROM memories WHERE id=$1 AND tenant_id=$2",
      [id, tenantId],
    );
    if (!result.rowCount) throw new NotFoundError("memory not found");
  }
  async retrieve(
    tenantId: string,
    query: string,
    scopes: Array<[string, string]>,
    limit = 8,
  ) {
    if (!scopes.length) return [];
    const rows = await this.list(tenantId, { enabled: true, limit: 1000 });
    const allowed = new Set(scopes.map(([type, id]) => `${type}:${id}`));
    const vector = hashEmbedding(query);
    return rows
      .filter((row) => allowed.has(`${row.scope_type}:${row.scope_id}`))
      .map((row) => ({
        ...row,
        similarity: cosine(vector, row.embedding_json),
      }))
      .sort((a, b) => b.similarity * b.confidence - a.similarity * a.confidence)
      .slice(0, limit);
  }
  async consolidateRun(
    tenantId: string,
    run: {
      id: string;
      agent_version_id: string;
      input_json: unknown;
      output_json: unknown;
    },
  ) {
    if (run.output_json === undefined || run.output_json === null)
      return undefined;
    const content =
      `Input: ${typeof run.input_json === "string" ? run.input_json : JSON.stringify(run.input_json)}\nOutcome: ${typeof run.output_json === "string" ? run.output_json : JSON.stringify(run.output_json)}`.slice(
        0,
        50_000,
      );
    return this.create(
      tenantId,
      {
        scope_type: "agent",
        scope_id: run.agent_version_id,
        kind: "episodic",
        content,
        confidence: 0.65,
      },
      { run_id: run.id },
    );
  }
}
